crates/turbo-tasks-fs/src/lib.rs

```diff
@@ -1196,6 +1198,69 @@ impl From<File> for FileContentVc {
     }
 }
 
+#[derive(Clone, Debug, Eq, PartialEq)]
+enum FileComparison {
+    Create,
+    Equal,
+    NotEqual,
+}
+
+impl FileContent {
+    /// Performs a comparison of self's data against a disk file's streamed
+    /// read.
+    async fn streaming_compare(&self, path: PathBuf) -> Result<FileComparison> {
+        let old_file = extract_disk_access(retry_future(|| fs::File::open(&path)).await, &path)?;
+        let Some(mut old_file) = old_file else {
+            return Ok(match self {
+                FileContent::NotFound => FileComparison::Equal,
+                _ => FileComparison::Create,
+            });
+        };
+        // We know old file exists, does the new file?
+        let FileContent::Content(new_file) = self else {
+            return Ok(FileComparison::NotEqual);
+        };
+
+        let old_meta = extract_disk_access(retry_future(|| old_file.metadata()).await, &path)?;
+        let Some(old_meta) = old_meta else {
+            // If we failed to get meta, then the old file has been deleted between the handle open.
+            // In which case, we just pretend the file never existed.
+            return Ok(FileComparison::Create);
+        };
+        // If the meta is different, we need to rewrite the file to update it.
+        if new_file.meta != old_meta.into() {
+            return Ok(FileComparison::NotEqual);
+        }
+
+        // So meta matches, and we have a file handle. Let's stream the contents to see
+        // if they match.
+        let mut new_contents = new_file.read();
+        let mut old_contents = AsyncBufReader::new(&mut old_file);
```

Comment:
Performance-wise it would probably be better to avoid the buffered reader here and just read directly from the old_file as much as possible. Probably requires a little bit more complicated compare logic to handle consumed positions.

Reply:
I think you're assuming reading from a file returns a buffer, and we're then copying from that buffer into our `BufReader` buffer. But that's not the case, the `read()` API requires you to maintain a buffer and that gets passed to the system to copy bytes into, reading doesn't return a system buffer.

So we can either maintain our own buffer in this loop, or we can reuse the buffer created as part of `BufReader`.

Reply:
I see :+1: